src/NuGet.Core/NuGet.Build.Tasks/GenerateTemporaryTargetAssembly2.cs

```diff
@@ -0,0 +1,523 @@
+
+//---------------------------------------------------------------------------
+//
+// <copyright file="GenerateTemporaryTargetAssembly.cs" company="Microsoft">
+//    Copyright (C) Microsoft Corporation.  All rights reserved.
+// </copyright>
+// 
+// Description: This is a MSBuild task which generates a temporary target assembly
+//              if current project contains a xaml file with local-type-reference.
+//
+//              It generates a temporary project file and then call build-engine 
+//              to compile it.
+//              
+//              The new project file will replace all the Reference Items with the 
+//              resolved ReferenctPath, add all the generated code file into Compile 
+//              Item list.
+//
+// History:
+//  05/10/05: weibz   Created.
+//  11/03/17: support NuGet PackageReference properly in temporary assembly by enabling .g.targets and .g.props to be
+//            properly included, despite the project file name change.
+//
+//---------------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Collections;
+
+using System.Globalization;
+using System.Diagnostics;
+using System.Reflection;
+using System.Resources;
+using System.Xml;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.BuildEngine;
+
+using MS.Utility;
+using MS.Internal.Tasks;
+
+// Since we disable PreSharp warnings in this file, PreSharp warning is unknown to C# compiler.
+// We first need to disable warnings about unknown message numbers and unknown pragmas.
+#pragma warning disable 1634, 1691
+
+namespace Microsoft.Build.Tasks.Windows
+{
+    #region GenerateTemporaryTargetAssembly2 Task class
+
+    /// <summary>
+    ///   This task is used to generate a temporary target assembly. It generates
+    ///   a temporary project file and then compile it.
+    /// 
+    ///   The generated project file is based on current project file, with below
+    ///   modification:
+    /// 
+    ///       A:  Add the generated code files (.g.cs) to Compile Item list.
+    ///       B:  Replace Reference Item list with ReferenctPath item list.
+    ///           So that it doesn't need to rerun time-consuming tatk 
+    ///           ResolveAssemblyReference (RAR) again.
+    /// 
+    /// </summary>
+    public sealed class GenerateTemporaryTargetAssembly2 : Task
```

Comment:
This Task wasn't previously OSS, was it? Do we have the necessary approvals to make it so?

Reply:
Good point. Yes, will have to deal with license side of this.